### PR TITLE
Python iRODS Client implementation of parallel transfer

### PR DESCRIPTION
**Pull request by d-w-moore**
This pull request addresses a foundational issue #232 ( and a smaller one in #233 ) in the Python iRODS-Client (PRC) and then provides a multithreaded data-transfer implementation for  the PRC that is similar (and comparable in performance) to the Parallel Transfer Engine integrated into the iRODS core code. The idea is to manage "N" connections to the server -- see iput and iget -N options -- each connection handling 1/N of the work of the (PUT|GET) operation. Because account credentials are available for reuse, this gets around the server having to encrypt the extra channels independent of, for example, the SSL encrypt capability inherent to the native iRODS connection.

### CODE REVIEW

**Comment by trel** [comment]
getting there...  codacy has some opinions.

**Comment by d-w-moore** [comment]
Yes , going to work on the
  - codacy suggestions
  - tests
  - diverting parallel to single channel under certain conditions
then run the whole test suite and let you know!

File: irods/manager/data_object_manager.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def _download(self, obj, local_path, auto_parallel = AUTO_SWITCH_PARALLEL, **opt
             parallel_xfer = False
             if auto_parallel and self.server_version >= [4,2,8]:
                 dobj_size = o.seek(0, os.SEEK_END)
-                if dobj_size > PARALLEL_THRESHOLD:
+                if dobj_size >= PARALLEL_THRESHOLD:
```
Comment (trel) [inline irods/manager/data_object_manager.py:50]: icommands have it the other way.   max for single buffer = 32MB...

https://github.com/irods/irods/blob/4-2-stable/server/api/src/rsDataObjOpen.cpp#L505

so remove the '=' to be consistent

**Comment by d-w-moore** [comment]
Tested and ready for merge, unless README is judge to need an update for the parallel transfer. (The demo script and progress bar niceties would be the only possibly compelling reason, as the data_objects manager put/get methods should transparently switch over to parallel mode when sizes are greater than 32MB. )

File: irods/test/admin_test.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 from irods.session import iRODSSession
 from irods.resource import iRODSResource
 import irods.test.helpers as helpers
+import  irods.keywords as kw
```
Comment (trel) [inline irods/test/admin_test.py:11]: whitespace

File: irods/test/data_obj_test.py [hunk 3]
```diff
@@ -27,13 +35,111 @@ def setUp(self):
...
-
     def tearDown(self):
         '''Remove test data and close connections
         '''
         self.coll.remove(recurse=True, force=True)
         self.sess.cleanup()
 
+    @staticmethod
+    def In_Memory_Stream():
+        return  io.BytesIO() if sys.version_info < (3,) else io.StringIO()
```
Comment (trel) [inline irods/test/data_obj_test.py:46]: whitespace

**Review by d-w-moore — COMMENTED** [review]
> consider as the environment variable... `ENABLE_PARALLEL_TRANSFER_FOR_428`.
> 
> leads the user/reader to ignore if not specifically interested in 4.2.8.

agreed

File: irods/manager/data_object_manager.py [hunk 4]
```diff
@@ -59,19 +82,35 @@ def get(self, path, file=None, **options):
...
+                        if not self.parallel_put( local_path, obj, num_threads = num_threads,
+                                                  target_resource_name = options.get(kw.RESC_NAME_KW,'') or
+                                                                         options.get(kw.DEST_RESC_NAME_KW,'')):
+                            raise RuntimeError("parallel put failed")
+                    except WrongServerVersion: pass
+                    else:  parallel_xfer = True
+            if not parallel_xfer:
+                file_size = f.seek(0,os.SEEK_SET)
+                for chunk in chunks(f, self.WRITE_BUFFER_SIZE):
+                    o.write(chunk)
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py]: done

File: irods/parallel.py [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+
+    replica_token = dobj_info.get("replica_token")
+    resc_hier = ( dobj_info.get("data_object_info") or {} ).get("resource_hierarchy")
+
+    logger.debug(json.dumps(dobj_info,indent=4))
+
+    num_threads = kwopt.pop( 'num_threads', None)
+    if num_threads is None: num_threads = int(kwopt.get('N','0'))
+
+    # TODO: allow part of file or data object to be transferred (`range_for_io' param)
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/parallel.py]: Yep ... not sure yet what semantics  it has to read/write a sub-range of an object (or several sub-ranges).   If it be defined as a "put"/"get" for example. Probably best leave it a TODO until these questions are answered

**Comment by d-w-moore** [comment]
This is working ideally now, and matches `iput` and `iget` for speed of parallelism on 4-2-stable when the environment variable  IRODS_VERSION_OVERRIDE is set to "4,2,9". (this won't be necessary after the version bump.) Also, the test suite is passing on 4-2-stable; modifications were needed on  a couple of tests, and those are included on this branch too.  

For those who desire to benchmark parallel transfer but are running 4.2.8 release of iRODS server, you can select an earlier [commit](https://github.com/irods/python-irodsclient/pull/236/commits/0f3840ee05c2fc74427c41c55ded1b0ef055739c) with which to test, but be aware that it cannot be considered as production quality. For one, acPostProcForPut() is called potentially  multiple times for a PUT (once per thread of a transfer).

File: PYTHON_install_caveats.rst [hunk 6]
```diff
@@ -1,64 +1,76 @@
...
-Python (2.x) and virtualenv for the iRODS administrator
-=======================================================
+==================================================
+Python, pip, virtualenv - caveats for use with PRC
+==================================================
 
 The Python iRODS Client (PRC) can be installed by a
-non-administrative user, either as a local install for
-the user (using ~/.local) or under a virtual environment.
+non-administrative user, either as a locally for
```
Comment (trel) [inline PYTHON_install_caveats.rst:6]: "as a locally" -> "locally"

**Comment by trel** [comment]
there are a couple merge conflicts - after those are cleaned up, we can squash down to clean commits and get every commit to have an issue number.

need to release 0.8.6 before we merge this (for what will be 0.9.0)

**Comment by d-w-moore** [comment]
I had planned on resolving those but was waiting to do everything at once ( test with 4-2-stable, assuming Alan's changes are pretty much all in.)

**Comment by trel** [comment]
yeah, these are minor, and i don't expect too much more interaction with the alan server code.

**Comment by d-w-moore** [comment]
> yeah, these are minor, and i don't expect too much more interaction with the alan server code.

ok will rebase and test this weekend

**Comment by trel** [comment]
codacy and then we start to squash.

**Comment by d-w-moore** [comment]
ok going to run tests again also.

File: irods/test/file_test.py [hunk 7]
```diff
@@ -76,20 +76,24 @@ def test_file_write(self):
...
-
-        # self.assertTrue(expr, msg)
-        f.close()
+        with obj.open('w+') as f:
+            f.write(self.write_str.encode('utf-8'))
+            f.seek(-6, 2)
+            f.write(self.write_str1.encode('utf-8'))
+
+            ## seek and re-read of just written bytes doesn't work in current 4-2-stable
+            ## (see https://github.com/irods/irods/issues/5315)
```
Comment (trel) [inline irods/test/file_test.py:85]: 5315 is now closed... so this can now be enabled i think

Comment (d-w-moore) [reply irods/test/file_test.py:85]: Yes , good point - will do that in upcoming testing !

**Pull request closed without merging**